Add Faster page (#2946)
neondatabase/website: content/faster/index.md

---
title: Faster is what we help you ship
subtitle: A page dedicated to teams shipping faster experiences faster on Neon.
enableTableOfContents: true
createdAt: '2025-02-12T00:00:00.000Z'
updatedOn: '2025-02-12T00:00:00.000Z'
---

> When we say "Ship faster with Postgres" -- faster is a noun and a verb. Our mission is not just about helping you build and ship softare faster. It's about helping you deliver that same feeling of faster to your customers.

Reviewer: ```suggestion
When we say "Ship faster with Postgres" — it's not just about helping you build and ship software faster. It's about helping you deliver that same feeling of **faster** to your customers.
```

## Open Source Faster

Peer behind the curtain and learn how faster gets shipped with these open source examples.

> 

Reviewer: maybe use em dashes? —

- **[Invenco](https://invenco.net)** – Handles high-traffic e-commerce spikes without needing to thinking about databases. [Details](https://neon.tech/blog/why-invenco-migrated-from-aurora-serverless-v2-to-neon)
- **[Branch Insurance](https://ourbranch.com)** – Builds on Serverless Postgres to deliver faster experience out-of-the-box. [Details](https://neon.tech/blog/branch-chose-neon-for-its-true-postgres-and-serverless-nature)
- **[Magic Circle](https://magiccircle.io)** – Handles 2M+ game sessions on Neon. [Details](https://neon.tech/blog/how-magic-circle-scaled-up-to-2m-games-with-cloudfare-and-neon)
- **[Cedalio](https://cedalio.com)** – One-database-per-client model with auto-suspend. [Details](https://neon.tech/blog/how-cedalio-uses-neon-for-an-efficient-development-workflow)
- **[BaseHub](https://basehub.com)** – Uses autoscaling to handle peak loads with zero manual tuning. [Details](https://neon.tech/blog/meet-basehub-developer-velocity-and-efficiency-right-down-to-the-database)
- **[Create.xyz](https://create.xyz)** – Spins up Postgres backends instantly for AI-generated apps. [Details](https://neon.tech/blog/from-idea-to-full-stack-app-in-one-conversation-with-create)

> - **[OpusFlow](https://opusflow.io)** – DB-per-tenant delivers fast, no-noisy-neighbor experience for energy customers. [Details](https://neon.tech/blog/how-opusflow-achieves-tenant-isolation-in-postgres-without-managing-servers)

Reviewer: ```suggestion
- **[OpusFlow](https://opusflow.io)** – DB-per-tenant delivers a fast, no-noisy-neighbor experience for energy customers. [Details](https://neon.tech/blog/how-opusflow-achieves-tenant-isolation-in-postgres-without-managing-servers)
```

- **[BeatGig](https://beatgig.com)** – Uses instant read replicas for fast analytics. [Details](https://neon.tech/blog/neon-read-replicas-in-the-wild-how-beatgig-uses-them)
- **[ketteQ](https://ketteq.com)** – Runs hundreds of forecast simulations in parallel with instant branches. [Details](https://neon.tech/blog/database-branching-for-postgres-with-neon)
- **[Topo.io](https://topo.io)** – AI GTM Platform outbuilds their competition with engineering grit and Neon. [Details](https://neon.tech/blog/why-topo-io-switched-from-amazon-rds-to-neon)
- **[Supergood.ai](https://supergood.ai)** – High-throughput workload powering AI-generated API's for any site. [Details](https://neon.tech/blog/how-supergood-unlocked-their-postgres-developer-productivity)
- **[Shepherd](https://shepherdinsurance.com)** – CI/CD test DBs spin up instantly via GitHub Actions. [Details](https://neon.tech/blog/adopting-neon-branching-in-ci-cd-pipelines-a-practical-story-by-shepherd)

Is your company using Neon to ship faster experiences faster? [Let us know in this GitHub Discussion](https://github.com/neondatabase/neon/discussions/10827) and we'll add you to the list!

## Faster Features

> We take pride in [owning our tenth of a second](https://x.com/tobi/status/1787139157078188180?lang=en). Here are some of the faster features we are particularly proud of.

Reviewer: I'm not easily groking this reference. Not sure what "our tenth of a second" refers to exactly, but I guess it refers to how we achieve "fast"

Maybe something like this: 
Took me a minute to grasp the idea behind the referencve. Maybe something to help bridge the reader:

We take pride in [owning our tenth of a second](https://x.com/tobi/status/1787139157078188180?lang=en). Great software isn’t just functional—it’s fast. At Neon, it’s something we refine and build into our service in countless small ways. Here are some of the faster features we’re particularly proud of.

You can read about the [architectural decisions for Neon](https://neon.tech/blog/architecture-decisions-in-neon) and even see [all of our engineering RFCs](https://github.com/neondatabase/neon/tree/main/docs/rfcs), too. If you're interested [we're hiring](https://neon.tech/careers).

## Faster Open Source Benchmarks

> Test faster yourself with these open-source third-party benchmarks.

Reviewer: ```suggestion
Test **faster** yourself with these open-source third-party benchmarks.
```

- https://db-latency.vercel.app/ - Open source (https://github.com/vercel-labs/function-database-latency) and written by Vercel

If you have an open-source benchmark that includes Neon [let us know here](#) we'd love to feature it on this page.

## Try it yourself, faster

It takes only ten seconds to

## How to ship faster

Shipping faster UX is a full-stack job. For databases, your real-world speed depends primarily on three things:

> 1. **Connection Approach:** Long-lived connections or connection for every query?

Reviewer: ```suggestion
1. **Connection Approach:** Long-lived connections or a connection for every query?
```

2. **Client-Database Proximity:** How close is your application (the client) to the database?
3. **Database Processing Time:** How much time is the database spending answering each query?

You can visualize the way each factor contributes like this:

<LatencyCalculator />

> **What about cold starts?** Databases on Neon can scale to zero when there are no active queries for a certain amount of time. We kept this topic separate below in [Cold Starts](#cold-starts) because _across the more than one million active databases on our platform, there are fewer than 5 cold starts per second._ Cold starts don't factor in to your real-world experience often enough to become an important consideration.

Reviewer: ```suggestion
**What about cold starts?** Databases on Neon can scale to zero when there are no active queries for a certain amount of time. We kept this topic separate below in [Cold Starts](#cold-starts) because _across the more than one million active databases on our platform, they occur very seldomly — there are fewer than 5 cold starts per second._ Cold starts don't factor in to your real-world experience often enough to become an important consideration.
```

3. Use an HTTP API via the [Neon Serverless Driver](https://neon.tech/docs/serverless/serverless-driver) - To reduce the number of roundtrips required to establish the connection.
   ![Serverless driver round-trips](https://neon.tech/_next/image?url=https%3A%2F%2Fneondatabase.wpengine.com%2Fwp-content%2Fuploads%2F2023%2F08%2FDevDays-GM-Quicker-2.jpg&w=3840&q=85&dpl=dpl_93LwG65BEFC73DGwgsAif219AFEd)
   Neon's [serverless driver](/docs/serverless-driver) and proxy have been optimized to reduce the number of roundtrips to the absolute minimum. To read more about how this works, see: [Quicker serverless Postgres connections](https://neon.tech/blog/quicker-serverless-postgres)
</Admonition>

### Putting app and database close together

Client-database proximity plays a major role in real-world database latency. Here are the different scenarios laid out from lowest to highest latency.

1. **App and database on same VM** - `1ms` to connect and `1ms-nms` to query. When you put your database and your app on the exact same machine, latency is measured in microseconds. _This setup is not possible on Neon because it is a distributed cloud platform._

> 2. **App and database in same region (datacenter) connecting over private network \-** 1ms to connect, 1ms to query \[Not exactly possible on Neon, you can connect through PrivateLink but it doesnt save you any time\]

Reviewer: ```suggestion
2. **App and database in same region (datacenter) connecting over private network \-** 1ms to connect, 1ms to query \[Not exactly possible on Neon, you can connect through PrivateLink but it doesnt save you any time\].
```

3. **App and database in same region (datacenter) connecting over public network** \- 3ms to connect, 1ms to query \- This is the fastest way to connect that is accessible to all Neon users. What you’re doing here is picking the same AWS or Azure region for your database and your application (be it serverless, like lambda functions on AWS, or serverful).
   1. You still use the publicly routable address of your database, but data is traveling 0 miles so it is very fast.
   2. This doesn’t mean your app has to be running on AWS directly, for example Vercel runs infra on AWS so picking the same Vercel and Neon region has the same effect.

> 4. **App and database in different region \-** Now we get into that compounding effect

Reviewer: ```suggestion
4. **App and database in different region \-** Now we get into that compounding effect.
```

### Minimizing time spent answering the query

As your business (and database) grows, connection and network transit latency remain static and database processing time becomes the most important factor to optimize. Here are some pointers:

- **Use connection pooling and autoscaling** – Use the built-in connection pooler to handle many client connections efficiently, and enable Neon’s autoscaling so the database can allocate sufficient RAM/CPU on demand.
- **Profile slow queries** – Identify which SQL statements are slow or resource-intensive by starting with the [Query History View](https://neon.tech/docs/introduction/monitor-query-history) in Neon.
  ![Neon query history tab](/docs/introduction/query_history.png)

>   You can dig deeper with `pg_stat_statements`](https://neon.tech/docs/extensions/pg_stat_statements)

Reviewer: ```suggestion
  You can dig deeper with [pg_stat_statements](https://neon.tech/docs/extensions/pg_stat_statements)
```

- **Reduce table and index bloat** – Reclaim wasted space and improve performance by eliminating bloat (accumulated dead rows) in tables and indexes. Schedule regular maintenance like `VACUUM` to remove dead tuples and use `REINDEX` on bloated indexes. You can also fine-tune autovacuum settings to keep bloat in check over time.
- **Leverage caching for reads** – Ensure frequently accessed data is served from memory instead of disk. Neon’s architecture extends Postgres’s shared memory buffers with a local file system cache, so aim for a high cache hit ratio. After installing the Neon extension, you can query the `neon_stat_file_cache` view to monitor how often data is read from cache versus storage and adjust your workload or memory allocation if needed.

**Further reading** – See [PostgreSQL query performance guide](https://neon.tech/docs/postgresql/query-performance) for in-depth explanations and tips.

## Definitions

To move fast, we must have a shared understanding of concepts and terms. Let's define some of the most commonly misinterpreted terms around how Neon works.

### Serverless

To us, serverless means:

- Instant Provisioning
- No server management
- Autoscaling

> - Usage-based pricing - _Your costs are directly tied to the resources your workload consumes—both compute and storage._

Reviewer: ```suggestion
- Usage-based pricing - _Your costs are directly tied to the resources your workload consumes—both compute and storage_
```
